pkg/retrieval/retrieval.go

```diff
@@ -112,134 +112,88 @@ func (s *Service) RetrieveChunk(ctx context.Context, addr swarm.Address, origin
-				ctx := tracing.WithContext(context.Background(), tracing.FromContext(topCtx))
+				ctx := tracing.WithContext(context.Background(), tracing.FromContext(ctx))
 
 				// get the tracing span
 				span, _, ctx := s.tracer.StartSpanFromContext(ctx, "retrieve-chunk", s.logger, opentracing.Tag{Key: "address", Value: addr.String()})
 				defer span.Finish()
 
-				peerAttempt++
-				s.metrics.PeerRequestCounter.Inc()
 				go func() {
-
 					// cancel the goroutine just with the timeout
-					ctx, cancel := context.WithTimeout(ctx, retrieveChunkTimeout)
+					ctx, cancel := context.WithTimeout(ctx, defaultTTL)
 					defer cancel()
-
-					chunk, peer, requested, err := s.retrieveChunk(ctx, addr, sp, origin)
-					select {
-					case resultC <- retrievalResult{
-						chunk:     chunk,
-						peer:      peer,
-						err:       err,
-						retrieved: requested,
-					}:
-					case <-ctx.Done():
-					}
-
+					s.retrieveChunk(ctx, resultChan, doneChan, addr, sp, origin)
 				}()
-			} else {
-				select {
-				case resultC <- retrievalResult{}:
-				case <-ctx.Done():
+
+				if allowedRetries <= 0 || allowedRetrieves <= 0 {
+					continue
 				}
-			}
 
-			select {
-			case <-ticker.C:
-				// break
-			case res := <-resultC:
+				timer.Reset(p90TTL)
+
+			case res := <-resultChan:
+
 				if errors.Is(res.err, topology.ErrNotFound) {
 					if sp.Saturated() {
 						// if no peer is available, and none skipped temporarily
 						s.logger.Tracef("retrieval: failed to get chunk %s", addr)
 						return nil, storage.ErrNotFound
-					} else {
-						// skip to next request round if any peers are only skipped temporarily
-						peerAttempt = maxSelects
 					}
 				}
+
 				if res.retrieved {
-					if res.err != nil {
-						if !res.peer.IsZero() {
-							s.logger.Debugf("retrieval: failed to get chunk %s from peer %s: %v", addr, res.peer, res.err)
-						}
-						peersResults++
-					} else {
+					if res.err == nil {
 						return res.chunk, nil
 					}
+					if !res.peer.IsZero() {
+						s.logger.Debugf("retrieval: failed to get chunk %s from peer %s: %v", addr, res.peer, res.err)
+					}
+				} else {
+					allowedRetrieves++
 				}
-			case <-ctx.Done():
-				s.logger.Tracef("retrieval: failed to get chunk %s: %v", addr, ctx.Err())
-				return nil, fmt.Errorf("retrieval: %w", ctx.Err())
-			}
-
-			// all results received, only successfully attempted requests are counted
-			if peersResults >= maxPeers {
-				s.logger.Tracef("retrieval: failed to get chunk %s", addr)
-				return nil, storage.ErrNotFound
-			}
 
-			// if we have not counted enough successful attempts but out of selection amount, reset
-			if peerAttempt >= maxSelects {
-
-				// forwarder gives up after first round of request attempts,
-				// rather than holding up the forwarding chain by waiting to be able to reserve downstream peer
-				if !origin {
+				if allowedRetries <= 0 || allowedRetrieves <= 0 {
```

Comment:
here again the same conditional like before, but this time causes an error to be propagated